access_request_handler.go

```diff
@@ -62,7 +62,7 @@ func (f *Fosite) NewAccessRequest(ctx context.Context, r *http.Request, session
 
 	if r.Method != "POST" {
 		return accessRequest, errors.WithStack(ErrInvalidRequest.WithDebug("HTTP method is not POST"))
-	} else if err := r.ParseForm(); err != nil {
+	} else if err := r.ParseMultipartForm(2 ^ 24); err != nil && err != http.ErrNotMultipart {
```

Comment:
`ParseMultipartForm` does not parse anything except the content type `multipart`, so the proper encoding has replaced with `multipart` as `ParseForm` has been replaced and if the form is not `multipart`, then the flow fails (so every oauth2 library :D ).

Parsing multipart should be a **fallback**, not the **default**!

Also, do we really need 16MB of data here? Smells like a possible DoS vulnerability!

Reply:
I've just seen this review. I've changed this to be 1^10 aka 1M. The other thing being
I don't know which version of go you are running but in 1.10 ParseMultipartForm first calls ParseForm and then it continues to do a multipart parse

Reply:
> I've just seen this review. I've changed this to be 2^20 aka 1M. 

Nice!

> I don't know which version of go you are running but in 1.10 ParseMultipartForm first calls ParseForm and then it continues to do a multipart parse

I think the tests run Go 1.8. We should definitely make this bc compatible with older versions of Go!

> Also in case you missed it. If you use the FormValue or PostFormValue of the request both of them call ParseMultipartForm with the default value of 32M!

I didn't know that. We are however only calling `PostForm` which requires a prior `ParseForm` (or it's nil). This led me to investigate though and it seems that `ParseForm` reads up to 10MB, so also quite a lot (not 32M but still). So 1MB is definitely fine here, and 16MB was probably too! Parsing explicitly only 1MB is even better though!

Reply:
I've checked the source of request.go in [1.8](https://github.com/golang/go/blob/555eb70db2f7ff4aa22768339425430c1dfdf6ac/src/net/http/request.go#L1203-L1239) and it seems that the ParseMultipartForm code is the same and it also calls ParseForm

Reply:
Ohh, I see, I didn't know that. I was looking at another line where multipart is ignored but apparently it's handled in the outer functions of that call. Thank you for clarifying that! Learned a ton about form parsing in this PR :)